### ISSUE TITLE
Use custom handler contracts? (possibly via convention)

### ISSUE THREAD

**Issue by artganify** [rebus-org/Rebus#420]
Our company previously used [NServiceBus](http://particular.net/nservicebus) for all our messaging, but being very pricey and bloated with various features we'd never need, we looked for a more lightweight, open alternative and decided to go with Rebus.

While everything works perfectly so far, we've got one big issue - the message handler contract. Reason: One of our main products features a plugin API which allows plugins to receive application events with a custom message handler contract, like this:

    public interface IMessageHandler<TMessage> {
        void Handle(TMessage message, IMessageContext context) {

        }
    }

While the `IMessageContext` is basically just a wrapper around the service bus to e.g. conveniently reply (`context.Reply(myreplymessage)`).

Previously with NServiceBus, we were able to hook our own message handlers because NServiceBus resolved handlers and sagas by configurable conventions, not relying on a specific interface.

We decided to write our own `IHandlerActivator` for Rebus, but then we also got to write our own `ITransactionContext`, write a custom pipeline step for also appending/passing the current `IBus` instance, creating handler proxies etc. etc. 

Long story short, is there an easier way to achieve this? Could this be a possible feature for the next version? Updating all our plugins as well as 3rd party provided ones (roughly 30-40) to use the message contract from Rebus as well as making Rebus a dependency of our plugin system would be an insane thing to do.

**Comment by mookid8000** [comment]
Hello there @artganify , did you figure out a way to solve your problem?

**Comment by mookid8000** [comment]
I was just goofing around with some code... wanted to see if I could somehow wrap all of Rebus and provide an API that would bridge handler invocation from Rebus' `IHandleMessages<TMessage>` to a custom `ICustomMessageHandler<TMessage>` interface

**Comment by mookid8000** [comment]
I've got something working now, actually... I'll put it up somewhere, then you can take a look

**Comment by mookid8000** [comment]
I put the code here: [CustomRebusHandlers](https://github.com/mookid8000/CustomRebusHandlers)

As you can see, I've created a custom handler interface [`ICustomMessageHandler<TMessage>`](https://github.com/mookid8000/CustomRebusHandlers/blob/master/CustomerRebusHandlers/Custom/ICustomMessageHandler.cs) which I have implemented [once to handle strings](https://github.com/mookid8000/CustomRebusHandlers/blob/master/CustomerRebusHandlers/Custom/Handlers/ExampleStringHandler.cs).

The trick lies in [`CustomHandlerWrapper<TMessage>`](https://github.com/mookid8000/CustomRebusHandlers/blob/master/CustomerRebusHandlers/Custom/Internals/CustomHandlerWrapper.cs) which delegates handler invocation to an injected handler of the custom type.

You can look at the code if you want - I think it's actually pretty interesting :grin:

**Comment by artganify** [comment]
@mookid8000 Hey there. Sorry for the late reply. Got caught up with other stuff. :)

Well, thank you very much for your time and setting up the comprehensive example. From a first glance, it looks like exactly what we were heading for. I'm not familiar with Castle though, do I get this right that `CustomHandlerWrapper` is registered as open generic and for every resolved `ICustomMessageHandler`, a `CustomHandlerWrapper` is being created?

**Comment by mookid8000** [comment]
Yes, that's how it works - and then I use property injection of a corresponding `IHandleMessages<TMessage>` in order to avoid errors when one cannot be resolved.

The "problem" is that Rebus will go to great lengths to be able to support [polymorphic dispatch](https://github.com/rebus-org/Rebus/wiki/Polymorphic-message-dispatch), therefore resolving e.g. `IHandleMessages<string>`, `IHandleMessages<IComparable>`, `IHandleMessages<ICloneable>`, `IHandleMessages<IEnumerable>`, `IHandleMessages<IConvertible>`, and ultimately `IHandleMessages<object>` in order to dispatch a `string` message.

You will most likely not have handlers for most of these, but the open generic will be resolved anyway. Therefore, the injected actual message handler must be optional somehow.
